### PR TITLE
Add in nix as a community support language.

### PR DESCRIPTION
**Pull request by matthewbauer**
This add a file called "lib/travis/build/nix.rb" that supports the "nix" language used in the "nix", "nixos", and "nixpkgs" projects.

- see nixos/nixpkgs#15749 for corresponding nixpkgs issue.
- see matthewbauer/nixpkgs@68ed840faeaa82986967afe5e070a281908e7d2a for example.

### CODE REVIEW

**Comment by domenkozar** [comment]
Me and @garbas are willing to co-maintain it.

**Comment by matthewbauer** [comment]
cc @BanzaiMan

File: lib/travis/build/script/nix.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# coding: utf-8
+# Maintained by
+#  - Domen Kožar
+#  - Rok Garbas
+#  - Matthew Bauer <mjbauer95@gmail.com>
```
Comment (BanzaiMan) [inline lib/travis/build/script/nix.rb]: This information needs to be displayed in the logs. See https://github.com/travis-ci/travis-build/blob/1bfb57d836416e5fda9d5cdc8986d681bbc4a9d7/lib/travis/build/script/csharp.rb#L20-L21 for an example.

Comment (domenkozar) [inline lib/travis/build/script/nix.rb]: @matthewbauer you can use my email domen@dev.si

Comment (matthewbauer) [inline lib/travis/build/script/nix.rb]: Ok cool.

cc @garbas

File: lib/travis/build/script/nix.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        def configure
+          super
+
+          # Set nix config dir and make config Hydra compatible
+          sh.cmd "sudo mkdir -p /etc/nix"
+          sh.cmd "echo 'build-max-jobs = 4' | sudo tee /etc/nix/nix.conf"
+
+          # Nix needs to be able to exec on /tmp on Linux
+          if config[:os] == 'linux' do
```
Comment (BanzaiMan) [inline lib/travis/build/script/nix.rb]: Drop `do`.

File: lib/travis/build/script/nix.rb (hunk 1 above)
Comment (garbas) [inline lib/travis/build/script/nix.rb]: @matthewbauer my email ... rok@garbas.si

**Comment by domenkozar** [comment]
@BanzaiMan @matthewbauer what's the status on this :)

**Comment by moretea** [comment]
What about adding support for caching `/nix` in here as well? [See this search query on the `travis-build` repo](https://github.com/travis-ci/travis-build/search?utf8=%E2%9C%93&q=%22data.cache%22&type=Code) for how this is implemented.

OTOH, the [docs](https://docs.travis-ci.com/user/caching/#Things-not-to-cache) explicitly ask not to cache packages.

File: lib/travis/build/script/nix.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# Maintained by
+#  - Domen Kožar   @domenkozar   <domen@dev.si>
+#  - Rok Garbas    @garbas       <rok@garbas.si>
+#  - Matthew Bauer @matthewbauer <mjbauer95@gmail.com>
```
Comment (BanzaiMan) [inline lib/travis/build/script/nix.rb]: Maintainers' names should appear in the logs. These comments will not.

**Comment by BanzaiMan** [comment]
We can test this tomorrow. (US East time)

**Comment by matthewbauer** [comment]
@moretea Travis stores caches as one big .tar.gz on AWS and Nix stores its cache as lots of small .tar.bz2 :)

Maybe we can figure out some better strategies for it but right now I think we should leave it up to Nix.

**Comment by BanzaiMan** [comment]
This has been deployed to staging.

File: lib/travis/build/script/nix.rb [hunk 4]
```diff
@@ -23,13 +23,14 @@ def configure
 
           # Set nix config dir and make config Hydra compatible
           sh.cmd "sudo mkdir -p /etc/nix"
-          sh.cmd "echo 'build-max-jobs = 4' | sudo tee /etc/nix/nix.conf"
+          sh.cmd "echo 'build-max-jobs = 4' | sudo tee /etc/nix/nix.conf > /dev/null"
```
Comment (domenkozar) [inline lib/travis/build/script/nix.rb]: No need to use `tee` then.

```
sudo echo 'build-max-jobs = 4' > /etc/nix/nix.conf
```

File: lib/travis/build/script/nix.rb [hunk 5]
```diff
@@ -23,13 +23,14 @@ def configure
...
 
           # Nix needs to be able to exec on /tmp on Linux
+          # This will emit an error in the container but
+          # it's still needed for "trusty" Linux.
```
Comment (domenkozar) [inline lib/travis/build/script/nix.rb]: We should detect if it's a container in that case.

Comment (matthewbauer) [inline lib/travis/build/script/nix.rb]: Definitely, but there doesn't seem to be a good way in Travis? cc @BanzaiMan

Comment (matthewbauer) [inline lib/travis/build/script/nix.rb]: Definitely, but there doesn't seem to be a good way in Travis? cc @BanzaiMan

File: lib/travis/build/script/nix.rb (hunk 4 above)
Comment (BanzaiMan) [inline lib/travis/build/script/nix.rb]: Redirection does not work with `sudo`.

Comment (matthewbauer) [inline lib/travis/build/script/nix.rb]: That will run "echo" as root but pipe to /etc/nix/nix.conf as normal user IIRC.

Comment (domenkozar) [inline lib/travis/build/script/nix.rb]:
```
$ sudo sh -c "ls > foo2"
$ ls -la foo2
-rw-r--r-- 1 root root 2945 Jul  1 16:01 foo2
```

Comment (matthewbauer) [inline lib/travis/build/script/nix.rb]:
```ruby
sh.cmd "sh -c 'echo \'build-max-jobs = 4\' > /etc/nix/nix.conf'"
```

**Comment by matthewbauer** [comment]
Here's the output of my test staging branch:

https://staging.travis-ci.org/matthewbauer/nixpkgs/builds/506011

I just pushed a couple of commits to make the logs a little nicer but everything looks good to me.

**Comment by BanzaiMan** [comment]
Deployed to staging again.

**Comment by domenkozar** [comment]
@matthewbauer @BanzaiMan where do we maintain documentation for this?

**Comment by BanzaiMan** [comment]
https://github.com/travis-ci/docs-travis-ci-com

**Comment by matthewbauer** [comment]
So I've got some sparse initial documentation at travis-ci/docs-travis-ci-com#659.

**Comment by matthewbauer** [comment]
:+1:

**Comment by domenkozar** [comment]
Does this mean we can start using `nix` language in production?

**Pull request merged**